src/ekore/harmonics/cache.py

```diff
@@ -0,0 +1,122 @@
+"""Caching (complicated) harmonic sums across :mod:`ekore`."""
+import numba as nb
+import numpy as np
+import numpy.typing as npt
+
+from . import w1, w2, w3, w4, w5
+
+# here a register of all possible functions
+S1 = 0  # = S_1(N)
+S2 = 1
+S3 = 2
+S4 = 3
+S5 = 4
+Sm1 = 5
+Sm2 = 6
+Sm3 = 7
+Sm4 = 8
+Sm5 = 9
+S21 = 10
+S2m1 = 11
+Sm21 = 12
+Sm2m1 = 13
+S31 = 14
+Sm31 = 15
+Sm22 = 16
+S211 = 17
+Sm211 = 18
+
+# this could also be S1h = S1(N/2)
+# the only requirement is that they are insubsequent order
+# and reset knows the maximum size (to fit them all)
+
+# this is a plain list holding the values
+@nb.njit(cache=True)
+def reset():
+    """Return the cache placeholder array."""
+    return np.full(19, np.nan, np.complex_)
+
+
+@nb.njit(cache=True)
+def get(key: int, cache: npt.ArrayLike, 
+        n: complex, is_singlet: bool) -> complex:
+    """Retrieve an element of the cache.
+
+    Parameters
+    ----------
+    key :
+        harmonic sum key
+    cache :
+        cache list holding all elements
+    n :
+        complex evaluation point
+    is_singlet :
+        symmetry factor: True for singlet like quantities 
+        (:math:`\eta=(-1)^N = 1`),
+        False for non-singlet like quantities 
+        (:math:`\eta=(-1)^N=-1`)
+
+    Returns
+    -------
+    complex :
+        requested harmonic sum evaluated at n
+
+    """
+    # Maybe improve error
+    if key < 0 or key > len(cache):
+        raise RuntimeError
+    # load the thing
+    s = cache[key]
+    # found? i.e. not NaN?
+    if not np.isnan(s):
+        return s
+    # compute it now ...
+    if key == S1:
+        s = w1.S1(n)
+    elif key == S2:
+        s = w2.S2(n)
+    elif key == S3:
+        s = w3.S3(n)
+    elif key == S4:
+        s = w4.S4(n)
+    elif key == S5:
+        s = w5.S5(n)
+    elif key == Sm1:
+        s = w1.Sm1(n, get(S1, cache, n), is_singlet)
```

Comment:
For example in this case, since `Sm1` depends on `S1` and one between `S1h` and `S1mh` (depending whether `is_singlet` is `True` or not) I would pass the whole cache, so that the function will check if the harmonic needed is  present, and if not, it will compute it.

Reply:
So if you see in the Sm1 parameters, cache is indeed being using for S1 to use an already computed S1 sum, however the Sm1 function has no input parameter option for the other two. That would require changing the wX files I think.

Reply:
Yes, what I'm saying is that I would change the input parameter of `Sm1`(and the same for `Sm2`, and `Sm3`) from `S1` to `cache` and then inside the function `w1.Sm1` I would call
```
if is_singlet:
    return c.get(c.S1h, cache, N, is_singlet) - S1
return c.get(c.S1mh, cache, N, is_singlet) - S1
```

Reply:
Obvioulsy you have to call
```
S1 = c.get(c.S1, cache, N, is_singlet)
```
at the beginning of the function

Reply:
It is a matter of design I would say where to compute the "helper s-es", e.g. `S1` for `Sm1` ... but we can decide to delegate it below (as suggested by @niclaurenti ) - especially in view of the arguments of `Sm22`